### PR TITLE
Refactoring ESP charge calculation

### PR DESCRIPTION
**Pull request by raimis**
- [x] Clean the fitting code
- [x] Separate into a separate class
- [x] Unit ests
  - [x] FakeQM
- [x] Documentation
  - [x] Doctests

This depends on #409 and #423.

### CODE REVIEW

**Comment by raimis** [comment]
@j3mdamas, this can be reviewed and merged after #409.

**Comment by j3mdamas** [comment]
I'll review this after #409 is successfully merged

**Comment by raimis** [comment]
@j3mdamas it is time for this PR!

File: htmd/data/test-param/benzamidine_esp_freeze_restart/parameters/CGenFF_2b6/B3LYP-6-31Gstar-vacuum/mol.psf [hunk 1]
```diff
@@ -17,10 +17,10 @@ PSF
       12 L    1    MOL  H5   HG61       0.0657262       0.0           0
       13 L    1    MOL  N1   N2P1             0.0       0.0           0
       14 L    1    MOL  N2   N2P1             0.0       0.0           0
-      15 L    1    MOL  H6   HGP2        0.240023       0.0           0
-      16 L    1    MOL  H7   HGP2        0.240023       0.0           0
-      17 L    1    MOL  H8   HGP2        0.240023       0.0           0
-      18 L    1    MOL  H9   HGP2        0.240023       0.0           0
+      15 L    1    MOL  H6   HGP2        0.327235       0.0           0
```
Comment (j3mdamas) [inline htmd/data/test-param/benzamidine_esp_freeze_restart/parameters/CGenFF_2b6/B3LYP-6-31Gstar-vacuum/mol.psf]: Why do only these charges change? Doesn't this mean the total charge of the molecule change? Seems weird.

File: .travis.yml [hunk 2]
```diff
@@ -63,6 +63,10 @@ install:
   - conda install $HOME/miniconda/conda-bld/*/htmd-data-[0-9]*.tar.bz2  -y -q
   - conda install $HOME/miniconda/conda-bld/*-64/htmd-[0-9]*.tar.bz2  -y -q
 
+  # TODO: move this to a proper place.
+  # TODO: there is "nlopt" on "acellera" channel, but it is broken or compiled without Python API.
+  - conda install -c conda-forge nlopt
```
Comment (j3mdamas) [inline .travis.yml]: I can do this, no problem

File: htmd/data/test-param/h2o2_esp/energies.txt [hunk 3]
```diff
@@ -5,7 +5,7 @@ Bond     : 0.09165582558617029
 Angle    : 0.13882333179780315
 Dihedral : 3.1999999910373385
 Improper : 0.0
-Electro  : 20.593075751152295
+Electro  : 5.7857768770164375
```
Comment (j3mdamas) [inline htmd/data/test-param/h2o2_esp/energies.txt]: What was going wrong before with ESP that now the fixed version provides such a different electrostatic energy?

Comment (j3mdamas) [reply htmd/data/test-param/h2o2_esp/energies.txt]: Hum, I see, the partial charges difference between oxygens and hydrogens went down.

File: htmd/parameterization/esp.py [hunk 4]
```diff
@@ -3,15 +3,82 @@
...
-    # A temporary class to collect ESP stuff
+    """
+    Electrostatic potential (ESP) charge fitting
+
+    Capabilities
+    ------------
+    - Generate points
+    - Consider equivalent atoms
+    - Impose total molecule charge
+    - Impose charge values bounds
```
Comment (j3mdamas) [inline htmd/parameterization/esp.py]: I'd rather write: "Impose boundaries for charge values".

File: htmd/parameterization/esp.py [hunk 5]
```diff
@@ -103,6 +190,169 @@ def _generate_points(molecule, vdw_radii=[1.4, 1.6, 1.8, 2.0, 2.2], density=10):
...
+
+    def _map_groups_to_atoms(self, group_charges):
+
+        charges = np.zeros(self.molecule.natoms)
+        for atom_group, group_charge in zip(self.molecule._equivalent_atom_groups, group_charges):
+            charges[atom_group] = group_charge
+
+        return charges
+
+    def _compute_constraint(self, group_charges, _):
```
Comment (j3mdamas) [inline htmd/parameterization/esp.py]: What's the `_` for as last argument?

File: htmd/parameterization/esp.py [hunk 6]
```diff
@@ -103,6 +190,169 @@ def _generate_points(molecule, vdw_radii=[1.4, 1.6, 1.8, 2.0, 2.2], density=10):
...
+        return lower_bounds, upper_bounds
+
+    def _compute_objective(self, group_charges, _):
+
+        qm_result = self.qm_results[0]
+
+        # Compute the reciprocal distances between ESP points and atomic charges
+        if self._reciprocal_distances is None:
+            distances = cdist(qm_result.esp_points, self.molecule.coords[:, :, 0])
+            distances *= 0.52917721067 # Convert Angstrom to a.u. (Bohr) (https://en.wikipedia.org/wiki/Bohr_radius)
```
Comment (j3mdamas) [inline htmd/parameterization/esp.py]: Some standard values use like on previous PRs? :)

File: htmd/qm/fake.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        result.errored = False
+        result.coords = self.molecule.coords
+        result.energy = 0 # TODO
+        result.dipole = self.molecule.getDipole()
+        result.esp_points = self.esp_points
+
+        # Compute ESP values
+        if self.esp_points is not None:
+            distances = cdist(result.esp_points, result.coords[:, :, 0]) # Angstrom
+            distances *= 0.52917721067  # Convert Angstrom to a.u. (Bohr) (https://en.wikipedia.org/wiki/Bohr_radius)
```
Comment (j3mdamas) [inline htmd/qm/fake.py]: Some more standard values here

**Comment by stefdoerr** [comment]
By the way for unit conversion I'd suggest using `from simtk import unit`
instead of hard-coding constants all over the code.
That or if you don't want to depend on openMM use pint (which we already
use in other places)

File: htmd/qm/fake.py (hunk 7 above)
Comment (raimis) [reply htmd/qm/fake.py]: Done!

File: htmd/parameterization/esp.py (hunk 6 above)
Comment (raimis) [reply htmd/parameterization/esp.py]: Done!

File: htmd/parameterization/esp.py (hunk 5 above)
Comment (raimis) [reply htmd/parameterization/esp.py]: For gradients, which aren't used in this case.

File: htmd/parameterization/esp.py (hunk 4 above)
Comment (raimis) [reply htmd/parameterization/esp.py]: Done!

File: htmd/data/test-param/h2o2_esp/energies.txt (hunk 3 above)
Comment (raimis) [reply htmd/data/test-param/h2o2_esp/energies.txt]: This is due to the unit conversion bug!

File: .travis.yml (hunk 2 above)
Comment (raimis) [reply .travis.yml]: Good luck!

File: htmd/data/test-param/benzamidine_esp_freeze_restart/parameters/CGenFF_2b6/B3LYP-6-31Gstar-vacuum/mol.psf (hunk 1 above)
Comment (raimis) [reply htmd/data/test-param/benzamidine_esp_freeze_restart/parameters/CGenFF_2b6/B3LYP-6-31Gstar-vacuum/mol.psf]: This test is still disable in this PR, so it doesn't matter.

**Comment by raimis** [comment]
@j3mdamas, I have addressed your comments!

File: htmd/data/test-param/benzamidine_esp_freeze_restart/parameters/CGenFF_2b6/B3LYP-6-31Gstar-vacuum/mol.psf (hunk 1 above)
Comment (j3mdamas) [reply htmd/data/test-param/benzamidine_esp_freeze_restart/parameters/CGenFF_2b6/B3LYP-6-31Gstar-vacuum/mol.psf]: OK, but we'll need to revise afterwards

File: htmd/parameterization/esp.py (hunk 5 above)
Comment (j3mdamas) [reply htmd/parameterization/esp.py]: then I don't understand what it is doing there

File: htmd/parameterization/esp.py [hunk 8]
```diff
@@ -103,6 +191,170 @@ def _generate_points(molecule, vdw_radii=[1.4, 1.6, 1.8, 2.0, 2.2], density=10):
...
+        # charge array. Note this also fixes the charges of the atoms in the
+        # same equivalency group.
+        for atom in self.fixed:
+            group = self.molecule._equivalent_group_by_atom[atom]
+            lower_bounds[group] = self.molecule.charge[atom]
+            upper_bounds[group] = self.molecule.charge[atom]
+
+        return lower_bounds, upper_bounds
+
+    def _compute_objective(self, group_charges, _):
```
Comment (j3mdamas) [inline htmd/parameterization/esp.py]: Same here. You really have to explain me the purpose of this parameter that is not used in the function

**Comment by j3mdamas** [comment]
I've changed the nlopt installation (integrating it into our conda). Shouldn't cause much trouble, but I got 3 failed builds. All qm-related, but different outcomes. Restarted, 1 passed, 2 failed. I'm restarting the failing ones now, but is there some inherent instability that we should be taking care instead? @raimis

**Comment by j3mdamas** [comment]
Oops, my mistake. The older nlopt packages on the acellera channel were of higher version and were being picked up xD

Let's see if it passes now

File: htmd/parameterization/ffmolecule.py [hunk 9]
```diff
@@ -219,137 +193,71 @@ def _fitDihedral_objective(self, x):
...
+        coords = self.coords[:, :, self.frame] - self.centreOfMass
 
-        # If the restraint relates to an H, set the lower bound to 0
-        for i in range(N):
-            if "H" == self.element[self._equivalent_atom_groups[i][0]]:
-                lb[i] = 0.001
+        dipole = np.zeros(4)
+        dipole[:3] = np.dot(self.charge, coords)
+        dipole[3] = np.linalg.norm(dipole[:3]) # Total dipole moment
+        dipole /= 0.20819434 # e * Ang --> Debye (https://en.wikipedia.org/wiki/Debye)
```
Comment (j3mdamas) [inline htmd/parameterization/ffmolecule.py]: There aren't Debyes in pint, but there is statCoulomb. Tomorrow I do this change. I'll also propose a PR to pint.

File: htmd/parameterization/esp.py (hunk 8 above)
Comment (raimis) [reply htmd/parameterization/esp.py]: This is a convention of *NLopt* (https://nlopt.readthedocs.io/en/latest/NLopt_Python_Reference/#Objective_function). It still pass `None` for gradient, even if it is not used.

**Comment by j3mdamas** [comment]
Ok perfect. Ready to merge.

On Oct 10, 2017 12:30, "Raimondas Galvelis" <notifications@github.com>
wrote:

*@raimis* commented on this pull request.
------------------------------

In htmd/parameterization/esp.py
<https://github.com/Acellera/htmd/pull/437#discussion_r143688486>:

> +        # If the restraint relates to an H, set the lower bound to 0
+        for i in range(self.ngroups):
+            if 'H' ==
self.molecule.element[self.molecule._equivalent_atom_groups[i][0]]:
+                lower_bounds[i] = 0.001
+
+        # Fix the charges of the specified atoms to those already set in the
+        # charge array. Note this also fixes the charges of the atoms in the
+        # same equivalency group.
+        for atom in self.fixed:
+            group = self.molecule._equivalent_group_by_atom[atom]
+            lower_bounds[group] = self.molecule.charge[atom]
+            upper_bounds[group] = self.molecule.charge[atom]
+
+        return lower_bounds, upper_bounds
+
+    def _compute_objective(self, group_charges, _):

This is a convention of *NLopt* (https://nlopt.readthedocs.io/
en/latest/NLopt_Python_Reference/#Objective_function). It still pass None
for gradient, even if it is not used.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Acellera/htmd/pull/437#discussion_r143688486>, or mute
the thread

**Pull request merged**